components/butterfly/src/rumor/dat_file.rs

```diff
@@ -111,99 +102,120 @@ impl DatFile {
             rumor_buf.resize(rumor_size as usize, 0);
             reader.read_exact(&mut rumor_buf)
                   .map_err(|err| Error::DatFileIO(self.path.clone(), err))?;
-            let rumor = Service::from_bytes(&rumor_buf)?;
-            server.insert_service(rumor);
             bytes_read += size_buf.len() as u64 + rumor_size;
+            op(&mut rumor_buf)?;
         }
 
-        debug!("Reading service-config rumors from {}",
+        Ok(())
+    }
+
+    fn read_rumor<F>(&mut self,
+                     rumor_type: &str,
+                     mut reader: &mut BufReader<File>,
+                     offset: u64,
+                     op: F)
+                     -> Result<()>
+        where F: FnMut(&mut Vec<u8>) -> Result<()>
+    {
+        debug!("Reading {} rumors from {}",
+               rumor_type,
                self.path().display());
-        bytes_read = 0;
-        loop {
-            if bytes_read >= self.header.service_config_len {
-                break;
-            }
-            reader.read_exact(&mut size_buf)
-                  .map_err(|err| Error::DatFileIO(self.path.clone(), err))?;
-            let rumor_size = LittleEndian::read_u64(&size_buf);
-            rumor_buf.resize(rumor_size as usize, 0);
-            reader.read_exact(&mut rumor_buf)
-                  .map_err(|err| Error::DatFileIO(self.path.clone(), err))?;
-            let rumor = ServiceConfig::from_bytes(&rumor_buf)?;
-            server.insert_service_config(rumor);
-            bytes_read += size_buf.len() as u64 + rumor_size;
-        }
+        self.read_and_process(&mut reader, offset, op)?;
+        Ok(())
+    }
 
-        debug!("Reading service-file rumors from {}", self.path().display());
-        bytes_read = 0;
-        loop {
-            if bytes_read >= self.header.service_file_len {
-                break;
-            }
-            reader.read_exact(&mut size_buf)
-                  .map_err(|err| Error::DatFileIO(self.path.clone(), err))?;
-            let rumor_size = LittleEndian::read_u64(&size_buf);
-            rumor_buf.resize(rumor_size as usize, 0);
-            reader.read_exact(&mut rumor_buf)
-                  .map_err(|err| Error::DatFileIO(self.path.clone(), err))?;
-            let rumor = ServiceFile::from_bytes(&rumor_buf)?;
-            server.insert_service_file(rumor);
-            bytes_read += size_buf.len() as u64 + rumor_size;
-        }
+    pub fn read_members<F>(&mut self, reader: &mut BufReader<File>, op: F) -> Result<()>
+        where F: FnMut(&mut Vec<u8>) -> Result<()>
+    {
+        self.read_rumor("membership", reader, self.header.member_len, op)
+    }
```

Comment:
I think this interface might be more ergonomic if it relied less on closures. For example, recasting `read_members` as:
```rust
    pub fn read_all_members(&mut self, reader: &mut BufReader<File>) -> Result<Vec<Membership>> {
        let mut members = Vec::new();
        self.read_rumor("membership", reader, self.header.member_len, |r| {
                members.push(Membership::from_bytes(&r)?);
                Ok(())
            })?;
        Ok(members)
    }
```
leads to the following (with some slight error handling/reporting changes):
```rust
        self.read_members(&mut reader, |r| {
                match Membership::from_bytes(&r) {
                    Ok(membership) => server.insert_member(membership.member, membership.health),
                    Err(err) => warn!("Error reading membership rumor from dat file, {}", err),
                }

                Ok(())
            })?;
```
becomes
```rust
        for Membership { member, health } in self.read_all_members(&mut reader)? {
            server.insert_member(member, health);
        }
```
…
```rust
    dat_file.read_members(&mut reader, |r| {
                match Membership::from_bytes(&r) {
                    Ok(m) => println!("{}", m),
                    Err(err) => warn!("Error reading membership rumor from dat file, {}", err),
                }

                Ok(())
            })?;
```
becomes
```rust
    for member in dat_file.read_all_members(&mut reader)? {
        println!("{}", member);
    }
```
…
```rust
    dat_file.read_members(&mut reader, |r| {
                match Membership::from_bytes(&r) {
                    Ok(_) => membership += 1,
                    Err(err) => warn!("Error reading membership rumor from dat file, {}", err),
                }

                Ok(())
            })?;
```
becomes
```rust
    membership += dat_file.read_all_members(&mut reader)?.len();
```

If you go this route, I think it might be possible to factor out the commonality in something like
```rust
pub fn read_all_members(&mut self, reader: &mut BufReader<File>) -> Result<Vec<Membership>>
```
to
```rust
pub fn read_all_rumors<T: protocol::Message>(&mut self, reader: &mut BufReader<File>) -> Result<Vec<T>>
```

Reply:
This turned out really well.  Thanks for the suggestion!